Packing with ContinuePackingAfterGeneratingNuspec=false should not fail when there are no errors

---
**Issue moved from dotnet/sdk#12292**
- Please respond to @aggieben.

---

_From @aggieben on Tuesday, June 30, 2020 7:41:03 PM_

Using the 5.0.0-preview.5.20279.10 SDK on Windows 10, the tooling fails to pack a nupkg for the project located here: https://github.com/aggieben/ConstrainedTypes (`master` branch).

I used the following command:
`dotnet pack -o pkg ConstrainedTypes.fsproj -v diag`.  The error message reported is:
```
"C:\Users\Ben\proj\oss\constrainedtypes\src\ConstrainedTypes\ConstrainedTypes.fsproj" (pack target) (1:7) ->
                   (GenerateNuspec target) ->
                     C:\Program Files\dotnet\sdk\5.0.100-preview.5.20279.10\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): error MSB4181: The "PackTask" task returned false but did not log an error. [C:\Users\Ben\proj\oss\constrainedtypes\src\ConstrainedTypes\ConstrainedTypes.fsproj]
```

The exact same command works just fine with SDK version 3.1.300.

Hi @aggieben, I tried the following steps to repro, but failed:
1. git clone https://github.com/aggieben/ConstrainedTypes
2. install dotnet SDK 5.0.0-preview.5.20279.10
3. go to constrainedtypes\src\ConstrainedTypes folder
4. run dotnet pack -o pkg ConstrainedTypes.fsproj -v diag
The error I encountered is:
```
Build FAILED.

                   "C:\repos\ConstrainedTypes\src\ConstrainedTypes\ConstrainedTypes.fsproj" (Restore target) (1) ->
                   "C:\repos\ConstrainedTypes\src\ConstrainedTypes\ConstrainedTypes.fsproj" (_GenerateRestoreGraphProjectEntry target) (1:5) ->
                   (PaketRestore target) ->
                     C:\repos\ConstrainedTypes\.paket\Paket.Restore.targets(171,5): error MSB3073: The command "dotnet paket restore" exited with code 1. [C:\repos\ConstrainedTypes\src\ConstrainedTypes\ConstrainedTypes.fsproj]

    0 Warning(s)
    1 Error(s)
```
I also tried to install dotnet SDK 3.1.400, but the result is the same as above.

Can you pls provide the repro steps if the above is not accurate? Thanks!

--- a/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs
+++ b/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs
@@ -120,9 +120,11 @@ namespace NuGet.Commands
         {
             outputPath = outputPath ?? GetOutputPath(builder, _packArgs, false, builder.Version);
             var successful = BuildPackage(builder, outputPath, symbolsPackage: false);
-            PackageArchiveReader packageArchiveReader = successful
-                ? new PackageArchiveReader(outputPath)
-                : null;
+            PackageArchiveReader packageArchiveReader = null;
+            if (successful && File.Exists(outputPath))
+            {
+                packageArchiveReader = new PackageArchiveReader(outputPath);
+            }
             return packageArchiveReader;
         }
 

--- a/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs
+++ b/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs
@@ -83,18 +83,17 @@ namespace NuGet.Commands
         /// <exception cref="PackagingException">If a core packaging validation fails.</exception>
         public bool RunPackageBuild()
         {
-            var reader = BuildPackage(Path.GetFullPath(Path.Combine(_packArgs.CurrentDirectory, _packArgs.Path)));
-            reader?.Dispose();
-            return reader != null;
+            var result = BuildPackage(Path.GetFullPath(Path.Combine(_packArgs.CurrentDirectory, _packArgs.Path)));
+            return result;
         }
 
         [Obsolete("Do not use this. Use RunPackageBuild() instead as it accounts for the effects of package analysis to the complete operation status.")]
         public void BuildPackage()
         {
-            BuildPackage(Path.GetFullPath(Path.Combine(_packArgs.CurrentDirectory, _packArgs.Path)))?.Dispose();
+            BuildPackage(Path.GetFullPath(Path.Combine(_packArgs.CurrentDirectory, _packArgs.Path)));
         }
 
-        private PackageArchiveReader BuildPackage(string path)
+        private bool BuildPackage(string path)
         {
             string extension = Path.GetExtension(path);
             if (extension.Equals(NuGetConstants.ManifestExtension, StringComparison.OrdinalIgnoreCase))
@@ -119,7 +118,12 @@ namespace NuGet.Commands
         [Obsolete("Do not use this. Use RunPackageBuild() instead as it accounts for the effects of package analysis to the complete operation status.")]
         public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPath = null)
         {
-            return BuildPackage(builder, outputPath, symbolsPackage: false);
+            outputPath = outputPath ?? GetOutputPath(builder, _packArgs, false, builder.Version);
+            var successful = BuildPackage(builder, outputPath, symbolsPackage: false);
+            PackageArchiveReader packageArchiveReader = successful
+                ? new PackageArchiveReader(outputPath)
+                : null;
+            return packageArchiveReader;
         }
 
         /// <summary>
@@ -132,7 +136,7 @@ namespace NuGet.Commands
         /// <param name="symbolsPackage">Whether this package is a symbols package. Symbols packages do not undergo validations.</param>
         /// <returns>A <see cref="PackageArchiveReader"/> if everything completed succesfully. Throws if a core package validation fails. Returns <see langword="null"/> if a validation rule got elevated from a warning to an error.</returns>
         /// <exception cref="PackagingException">If a core packaging validation fails.</exception>
-        private PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPath = null, bool symbolsPackage = false)
+        private bool BuildPackage(PackageBuilder builder, string outputPath = null, bool symbolsPackage = false)
         {
             outputPath = outputPath ?? GetOutputPath(builder, _packArgs, false, builder.Version);
             Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
@@ -160,44 +164,46 @@ namespace NuGet.Commands
                 PrintVerbose(outputPath, builder);
             }
 
-            var package = new PackageArchiveReader(outputPath);
-
-            if (package != null && !_packArgs.NoPackageAnalysis && !symbolsPackage)
+            using (var package = new PackageArchiveReader(outputPath))
             {
-                AnalyzePackage(package);
-                if (_packArgs.Logger is PackCollectorLogger collectorLogger)
+
+                if (package != null && !_packArgs.NoPackageAnalysis && !symbolsPackage)
                 {
-                    if (collectorLogger.Errors.Any(e => e.Level == LogLevel.Error))
+                    AnalyzePackage(package);
+                    if (_packArgs.Logger is PackCollectorLogger collectorLogger)
                     {
-                        package.Dispose();
-                        if (!isExistingPackage && File.Exists(outputPath))
+                        if (collectorLogger.Errors.Any(e => e.Level == LogLevel.Error))
                         {
-                            File.Delete(outputPath);
+                            package.Dispose();
+                            if (!isExistingPackage && File.Exists(outputPath))
+                            {
+                                File.Delete(outputPath);
+                            }
+                            return false;
                         }
-                        return null;
                     }
                 }
-            }
 
-            if (_packArgs.InstallPackageToOutputPath)
-            {
+                if (_packArgs.InstallPackageToOutputPath)
+                {
+                    _packArgs.Logger.Log(
+                        PackagingLogMessage.CreateMessage(
+                            string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.Log_PackageCommandInstallPackageToOutputPath, "Package", outputPath),
+                            LogLevel.Minimal));
+
+                    WriteResolvedNuSpecToPackageOutputDirectory(builder);
+                    WriteSHA512PackageHash(builder);
+                }
+
                 _packArgs.Logger.Log(
                     PackagingLogMessage.CreateMessage(
-                        string.Format(
-                            CultureInfo.CurrentCulture,
-                            Strings.Log_PackageCommandInstallPackageToOutputPath, "Package", outputPath),
+                        string.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandSuccess, outputPath),
                         LogLevel.Minimal));
-
-                WriteResolvedNuSpecToPackageOutputDirectory(builder);
-                WriteSHA512PackageHash(builder);
             }
 
-            _packArgs.Logger.Log(
-                PackagingLogMessage.CreateMessage(
-                    string.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandSuccess, outputPath),
-                    LogLevel.Minimal));
-
-            return package;
+            return true;
         }
 
         /// <summary>
@@ -645,18 +651,18 @@ namespace NuGet.Commands
             }
         }
 
-        private PackageArchiveReader BuildFromNuspec(string path)
+        private bool BuildFromNuspec(string path)
         {
             PackageBuilder packageBuilder = CreatePackageBuilderFromNuspec(path);
 
-            PackageArchiveReader packageArchiveReader;
+            bool successful;
 
             InitCommonPackageBuilderProperties(packageBuilder);
 
             if (_packArgs.InstallPackageToOutputPath)
             {
                 string outputPath = GetOutputPath(packageBuilder, _packArgs);
-                packageArchiveReader = BuildPackage(packageBuilder, outputPath: outputPath, symbolsPackage: false);
+                successful = BuildPackage(packageBuilder, outputPath: outputPath, symbolsPackage: false);
             }
             else
             {
@@ -677,15 +683,15 @@ namespace NuGet.Commands
                     }
                 }
 
-                packageArchiveReader = BuildPackage(packageBuilder, symbolsPackage: false);
+                successful = BuildPackage(packageBuilder, symbolsPackage: false);
 
                 if (_packArgs.Symbols)
                 {
-                    BuildSymbolsPackage(path);
+                    successful = successful && BuildSymbolsPackage(path);
                 }
             }
 
-            return packageArchiveReader;
+            return successful;
         }
 
         private PackageBuilder CreatePackageBuilderFromNuspec(string path)
@@ -723,7 +729,7 @@ namespace NuGet.Commands
                 _packArgs.Deterministic);
         }
 
-        private PackageArchiveReader BuildFromProjectFile(string path)
+        private bool BuildFromProjectFile(string path)
         {
             // PackTargetArgs is only set for dotnet.exe pack code path, hence the check.
             if ((string.IsNullOrEmpty(_packArgs.MsBuildDirectory?.Value)
@@ -775,28 +781,24 @@ namespace NuGet.Commands
 
             InitCommonPackageBuilderProperties(mainPackageBuilder);
 
+            bool successful = true;
             // Build the main package
             if (GenerateNugetPackage)
             {
-                PackageArchiveReader packageArchiveReader;
                 if (_packArgs.InstallPackageToOutputPath)
                 {
                     string outputPath = GetOutputPath(mainPackageBuilder, _packArgs);
-                    packageArchiveReader = BuildPackage(mainPackageBuilder, outputPath: outputPath, symbolsPackage: false);
+                    successful = BuildPackage(mainPackageBuilder, outputPath: outputPath, symbolsPackage: false);
                 }
                 else
                 {
-                    packageArchiveReader = BuildPackage(mainPackageBuilder, symbolsPackage: false);
+                    successful= BuildPackage(mainPackageBuilder, symbolsPackage: false);
                 }
 
                 // If we're excluding symbols then do nothing else
                 if (!_packArgs.Symbols || _packArgs.InstallPackageToOutputPath)
                 {
-                    return packageArchiveReader;
-                }
-                else
-                {
-                    packageArchiveReader?.Dispose();
+                    return successful;
                 }
             }
 
@@ -832,11 +834,11 @@ namespace NuGet.Commands
 
                 if (GenerateNugetPackage)
                 {
-                    return BuildPackage(symbolsBuilder, outputPath, symbolsPackage: true);
+                    successful = successful && BuildPackage(symbolsBuilder, outputPath, symbolsPackage: true);
                 }
             }
 
-            return null;
+            return successful;
         }
 
         private void CheckForUnsupportedFrameworks(PackageBuilder builder)
@@ -985,7 +987,7 @@ namespace NuGet.Commands
             return path;
         }
 
-        private void BuildSymbolsPackage(string path)
+        private bool BuildSymbolsPackage(string path)
         {
             PackageBuilder symbolsBuilder = CreatePackageBuilderFromNuspec(path);
             if (_packArgs.SymbolPackageFormat == SymbolPackageFormat.Snupkg) // Snupkgs can only have 1 PackageType. 
@@ -1014,7 +1016,7 @@ namespace NuGet.Commands
             string outputPath = GetOutputPath(symbolsBuilder, _packArgs, symbols: true);
 
             InitCommonPackageBuilderProperties(symbolsBuilder);
-            BuildPackage(symbolsBuilder, outputPath, symbolsPackage: false)?.Dispose();
+            return BuildPackage(symbolsBuilder, outputPath, symbolsPackage: false);
         }
 
         internal void AnalyzePackage(PackageArchiveReader package)

--- a/test/NuGet.Core.Tests/NuGet.Commands.Test/PackCommandRunnerTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Commands.Test/PackCommandRunnerTests.cs
@@ -7,7 +7,9 @@ using System.IO;
 using System.IO.Compression;
 using System.Linq;
 using NuGet.Common;
+using NuGet.Packaging;
 using NuGet.Test.Utility;
+using NuGet.Versioning;
 using Xunit;
 
 namespace NuGet.Commands.Test
@@ -15,7 +17,7 @@ namespace NuGet.Commands.Test
     public class PackCommandRunnerTests
     {
         [Fact]
-        public void BuildPackage_WithDefaultExcludes_ExcludesDefaultExcludes()
+        public void RunPackageBuild_WithDefaultExcludes_ExcludesDefaultExcludes()
         {
             using (var test = DefaultExclusionsTest.Create())
             {
@@ -43,6 +45,44 @@ namespace NuGet.Commands.Test
                         Assert.Equal(1, package.Entries.Count(entry => entry.Name == expectedEntryName));
                     }
                 }
+            }
+        }
+
+        [Fact]
+        public void RunPackageBuild_WithGenerateNugetPackageFalse_ReturnsTrue()
+        {
+            using (var testDirectory = TestDirectory.Create())
+            {
+                var args = new PackArgs()
+                {
+                    CurrentDirectory = testDirectory.Path,
+                    Logger = NullLogger.Instance,
+                    PackTargetArgs = new MSBuildPackTargetArgs()
+                    {
+                        NuspecOutputPath = Path.Combine(testDirectory.Path, "obj", "Debug"),
+                        ContentFiles = new Dictionary<string, IEnumerable<ContentMetadata>>()
+                    },
+                    Path = string.Empty,
+                    Exclude = Array.Empty<string>()
+                };
+                var packageBuilder = new PackageBuilder()
+                {
+                    Id = "test",
+                    Version = new NuGetVersion(1, 0, 0),
+                    Description = "Testing PackCommandRunner.GenerateNugetPackage = false"
+                };
+                packageBuilder.Authors.Add("tester");
+
+                var runner = new PackCommandRunner(args, MSBuildProjectFactory.ProjectCreator, packageBuilder);
+                runner.GenerateNugetPackage = false;
+
+                // Act
+                var actual = runner.RunPackageBuild();
+
+                // Assert
+                Assert.True(actual, "PackCommandRunner.RunPackageBuild was not successful");
+                var expectedNuspecPath = Path.Combine(args.PackTargetArgs.NuspecOutputPath, "test.1.0.0.nuspec");
+                Assert.True(File.Exists(expectedNuspecPath), "nuspec file does not exist");
             }
         }
 

